DataGrid calculateGroupValue: Add text and examples for docissues (#1507)
DevExpress/devextreme-documentation: api-reference/_hidden/dxDataGridColumn/calculateGroupValue.md

---
---
##### shortDescription
A field name by which to group column cells or a function that returns a field name or a calculated value for this operation. This value is also used to sort **DataGrid**'s groups.

##### param(rowData): Object
The current row's data.

##### return: any
The calculated value used to create and sort groups.

---

> The **DataGrid** groups its columns by the grouping column. If you want to use custom grouping, change values of this column in the **calculateGroupValue** function. As a result, group cells' headers will display the changed values that may include extra information. To remove it, specify a custom template in the [groupCellTemplate](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellTemplate) option ([groupCellRender](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellRender) in React). Refer to this  <a href="https://github.com/DevExpress-Examples/DataGrid---How-to-apply-custom-sorting-to-a-grouped-column" target="_blank">github repository</a> for an example. 

Reviewer: ```suggestion
The **DataGrid** groups its columns by the grouping column. If you want to use custom grouping, change values of this column in the **calculateGroupValue** function. As a result, group cells' headers will display the changed values that may include extra information. To remove it, specify a custom template in the [groupCellTemplate](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellTemplate) option ([groupCellRender](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellRender) in React). Refer to this <a href="https://github.com/DevExpress-Examples/DataGrid---How-to-apply-custom-sorting-to-a-grouped-column" target="_blank">GitHub repository</a> for an example. 
```

---
---
##### shortDescription

> A field name or a function that returns a field name or value by which to group cells. This value is also used to sort **DataGrid**'s groups.

Reviewer 2: "A field name" "that returns a field name"?  Can you explain?

Author: calculateGroupValue specifies a field name OR a function. This function should return a field name or value by which to group column cells.
Here is another variant: "A field name by which to group column cells or a function that returns a field name or a calculated value to perform this operation."
Is it better?

Reviewer 2: A field name by which to group column cells or a function that returns a field name or a calculated value for this operation?

##### param(rowData): Object
The current row's data.

##### return: any

> The calculated value to be used for grouping and group sorting.

Reviewer 2: ```suggestion
The calculated value used to group and sort columns.
```

Reviewer 2: Why "group sorting"?

Author: "group sorting" because the widget sorts groups, not grouped rows. A group is a set or rows.
In this picture the groups are circled:
![image](https://user-images.githubusercontent.com/66619750/100236427-b53bb900-2f3e-11eb-8038-658808963e0f.png)

Can the sentence be "The calculated value by which to create and sort groups."?

Reviewer 2: The calculated value used to create and sort groups.

---

> The **DataGrid** groups its columns by the grouping column. If you want to use custom grouping, change values of this column in the **calculateGroupValue** function. As a result, group cells' headers will display the changed values that may include extra information. To remove it, specify a custom template in the [groupCellTemplate](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellTemplate) option ([groupCellRender](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellRender) in React). The example below uses this concept to display the *"California"* group always on top regardless of the sort order:

Reviewer 2: I'm not sure why you add "As a result, group cells' headers will display the changed values that may include extra information". Can you explain?

Reviewer: Well, this is my suggestion so I'll try to make it a bit more clear. When users implement the calculateGroupValue, the returned value(that we will later pass to group cells headers and so on) may contain information that was not intended to be in the headers. 
In the next sentence, we show a way to avoid this.

---

> ##### jQuery

Reviewer: @pavel-2020 Just a note - we should replace this code with a link to github. The last sentence from the previous paragraph (The example below uses this concept to display the *"California"* group always on top regardless of the sort order:) should also be replaced with something like 'Review `[a project on GitHub](link)` that illustrates this scenario.'

---
---
##### shortDescription

> A field name by which to group cells or a function that returns a field name or a value to perform this operation. This value is also used to sort **DataGrid**'s groups.

Reviewer: 1. 
>A field name by which to group cells or a function that returns a field name or a value to perform this operation.

I would mix your structure and the initial one: A field name or a function that returns a field's name or value by which to group cells.
2. This value is also used to sort **DataGrid**'s groups - just to be clear - only value, not name, right?

Author: 1) Will apply it.
2) Yes, only value.

##### param(rowData): Object
The current row's data.

##### return: any
The calculated value to be used for grouping and group sorting.

---

> If you want to use modified values of a grouping column, assign

Reviewer: Seems broken

> By default, [grouping](/concepts/05%20Widgets/DataGrid/45%20Grouping '/Documentation/Guide/Widgets/DataGrid/Grouping/') is conducted using the exact values that the grouping column contains. If you want to use calculated values, implement the **calculateGroupValue** function. As a consequence, group cells' headers may display extra or missing information. To remove or add it, implement the [groupCellTemplate](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellTemplate) option ([groupCellRender](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellRender) in React). The example below displays the *"California"* group always on top regardless of the sort order. For this, **calculateGroupValue** and **groupCellTemplate** are used:

Reviewer: >By default, [grouping](/concepts/05%20Widgets/DataGrid/45%20Grouping '/Documentation/Guide/Widgets/DataGrid/Grouping/') is conducted using the exact values that the grouping column contains.

We should rewrite this old sentence to avoid 'by default'. How about the following: 'The **DataGrid** groups columns by the exact values that the grouping column contains. 

>As a consequence, group cells' headers may display extra or missing information.

I think that headers cannot display **missing** information. That's why it is missing, isn't it?
I would also replace 'As a consequence' with something less complicated - maybe 'After that' or similar.
.
>To remove or add it, implement the [groupCellTemplate](/Documentation/ApiReference/UI_Widgets/dxDataGrid/Configuration/columns/#groupCellTemplate) option

To display the information correctly, implement...

>For this, **calculateGroupValue** and **groupCellTemplate** are used

Let's get rid of passive. We may use the first person plural instead.

                        let sortValue = rowData.State === "California" ? this.sortOrder == "asc" ? "aaa" : "zzz" : rowData.State;
                        let displayValue = rowData.State;
                        return sortValue + ";" + displayValue;
                    },
                    groupCellTemplate: function(element, options) {
                        let column = options.column;
                        let displayValue = options.value.split(";")[1];
                        element.text(column.caption + ": " + displayValue);
                    }
            // ...
            ]
        });
    });

##### Angular
    
    <!--HTML-->
    <dx-data-grid ...>
        <!-- ... -->
        <dxi-column dataField="City"></dxi-column>
        <dxi-column
            dataField="State"
            [groupIndex]="0"
            [calculateGroupValue]="calculateGroupValue"
            groupCellTemplate="groupCellTemplate">
        </dxi-column>

        <div *dxTemplate="let data of 'groupCellTemplate'">
            <div>
                {{data.column.caption}}: {{data.value.split(";")[1]}}
            </div>
        </div>
    </dx-data-grid>

>     <!--TypeScript-->

Reviewer: We should use the following combination of tabs: app.component.html, app.component.ts, and app.module.ts with an optional app.service.ts (that isn't required here).
